<a href="https://colab.research.google.com/github/Argentan/DMA_LAB2/blob/master/tutoriales/02_intro_python2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Importación de Librerías

Importar una librería entera


In [1]:
import random

Puedo asignar un nombre corto para facilitar la llamada de funciones

In [2]:
import pandas as pd

Importar solo una función de una librería

In [3]:
from random import sample

Si no tengo instalada una librería puedo hacerla de varias maneras, la más simple usando **pip** 

El "**!**" lo usamos en Jupyter Notebooks para indicarle que es un comando de consola

In [4]:
!pip install seaborn

Defaulting to user installation because normal site-packages is not writeable



[notice] A new release of pip is available: 23.0.1 -> 23.1.2
[notice] To update, run: python.exe -m pip install --upgrade pip


Otra alternativa es con **apt-get**, por ejemplo, para esta librería que nos ayuda a leer archivos de compresión 7zip

In [5]:
# https://pypi.python.org/pypi/libarchive
!apt-get -qq install -y libarchive-dev && pip install -U libarchive

'apt-get' is not recognized as an internal or external command,
operable program or batch file.


También puedo ver la versión de la Librería que tengo usando **pip show**

In [6]:
!pip show tensorflow

Name: tensorflow
Version: 2.12.0
Summary: TensorFlow is an open source machine learning framework for everyone.
Home-page: https://www.tensorflow.org/
Author: Google Inc.
Author-email: packages@tensorflow.org
License: Apache 2.0
Location: C:\Program Files\Python311\Lib\site-packages
Requires: tensorflow-intel
Required-by: 


También puedo indicarle una versión específica de la librería si lo requiero

In [7]:
!pip install tensorflow

Defaulting to user installation because normal site-packages is not writeable



[notice] A new release of pip is available: 23.0.1 -> 23.1.2
[notice] To update, run: python.exe -m pip install --upgrade pip


# Estructuras de control

## Comparadores lógicos

| Operador | Acción |
| -------- | ------ |
| ==       | igual  |
| > =       | mayor o igual |
| >        | mayor estricto |
| < =      | menor o igual |
| <     | menor estricto |
| ! = | desigual|
| or | o |
| and | y |


In [8]:
1 == 1

True

In [9]:
1 == 2

False

In [10]:
1 == 1.

True

In [11]:
"hola" == "hola"

True

In [12]:
"hola" == "Hola"

False

In [13]:
"hola" != "Hola"

True

In [14]:
(1 == 1) and (2 == 2)

True

In [15]:
(1 == 1) and (2 != 2)

False

In [16]:
(1 == 1) or (2 != 2)

True

In [17]:
not ((1 == 1) or (2 != 2))

False

In [18]:
not (1 == 1) or (2 == 2)

True

### El operador "not in"

In [19]:
'Juan' in ['Pedro', 'Lucia', 'Juan', 'Maria']

True

In [20]:
'Juan' not in ['Pedro', 'Lucia', 'Juan', 'Maria']

False

In [21]:
'Julieta' not in ['Pedro', 'Lucia', 'Juan', 'Maria']

True

### Asignación con operador aritmético

| Operador | Equivalente |
| -------- | ----------- |
| `a += 1` | `a = a + 1` |
| `a -= 1` | `a = a - 1` |
| `a *= 1` | `a = a * 1` |
| `a /= 1` | `a = a / 1` |
| `a %= 1` | `a = a % 1` |

In [22]:
a = 'Hello'
a += ' world!'
a

'Hello world!'

In [23]:
a = 2
print(a)

a *= 2
print(a)

a *= 2
print(a)

2
4
8


## Condicionales

In [24]:
persona = "Juan"
if (persona == "Juan"):
    print("encontrado")

encontrado


In [25]:
if (persona == "Pedro"):
    print("encontrado")
else:
    print("no encontrado")

no encontrado


In [26]:
if (persona == "Pedro"):
    print("econtrado Pedro")
elif (persona == "Juan"): 
    print("econtrado Juan")
else:
    print("no encontrado")

econtrado Juan


## Loops

### Loops usando For

In [27]:
for a in [1, 2, 3]:
    print(a)

1
2
3


In [28]:
lista = [1, 2, 3]
for a in lista:
    print(a)

1
2
3


In [29]:
diccionario = {"v1": 1, "v2": 2, "v3": 3}
for a in diccionario:
    print(a)

v1
v2
v3


In [30]:
diccionario = {"v1": 1, "v2": 2, "v3": 3}
for a in diccionario.keys():
    print(a, diccionario[a])

v1 1
v2 2
v3 3


In [31]:
diccionario.items()

dict_items([('v1', 1), ('v2', 2), ('v3', 3)])

In [32]:
diccionario = {"v1": 1, "v2": 2, "v3": 3}
for k, v in diccionario.items():
    print(f"llave: {k}, valor: {v}")

llave: v1, valor: 1
llave: v2, valor: 2
llave: v3, valor: 3


In [33]:
v1, v2 = 1, 2
print(v1)
print(v2)

1
2


In [34]:
lista = ["a", "b", "c"]
for indice in range(len(lista)):
    print(f"el indice {indice} tienen un valor {lista[indice]}")

el indice 0 tienen un valor a
el indice 1 tienen un valor b
el indice 2 tienen un valor c


In [35]:
range(len(lista))

range(0, 3)

In [36]:
for indice, elemento in enumerate(lista):
    print(f"el indice {indice} tienen un valor {elemento}")

el indice 0 tienen un valor a
el indice 1 tienen un valor b
el indice 2 tienen un valor c


### Loops usando While 

In [37]:
from random import random

In [39]:
cent = 0
print(cent)
while cent < 0.7:
    cent = random()
    print("nuevo cent:", cent)

0
nuevo cent: 0.6797178116204797
nuevo cent: 0.6645325788935312
nuevo cent: 0.5201982657517509
nuevo cent: 0.1813787990681185
nuevo cent: 0.12611149770139385
nuevo cent: 0.17206895298995117
nuevo cent: 0.3325594930892606
nuevo cent: 0.8574715141365155


### Loop de múltiples Listas usando zip()


In [40]:
name = ['Pedro', 'Lucia', 'Juan', 'Maria']
age = [8, 9, 3, 5]
for n, a in zip(name, age):
  print('La calificación de {} fue {} sobre 10'.format(n, a))

La calificación de Pedro fue 8 sobre 10
La calificación de Lucia fue 9 sobre 10
La calificación de Juan fue 3 sobre 10
La calificación de Maria fue 5 sobre 10


### List Comprehension vs. For Loops

In [41]:
%%timeit
lista1 = []
for a in range(10000):
    b = a / 2 + 10
    lista1.append(b)

1.18 ms ± 81.7 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


In [42]:
%%timeit
lista2 = [a / 2 + 10 for a in range(10000)]

948 µs ± 42.4 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


## Referencia y Copia de Objetos

In [44]:
lista1 = []
for a in range(10):
    b = a / 2 + 10
    lista1.append(b)

In [45]:
lista2 = [a / 2 + 10 for a in range(10)]

In [46]:
print(lista1)
print(lista2)

[10.0, 10.5, 11.0, 11.5, 12.0, 12.5, 13.0, 13.5, 14.0, 14.5]
[10.0, 10.5, 11.0, 11.5, 12.0, 12.5, 13.0, 13.5, 14.0, 14.5]


In [47]:
print(lista1 is lista2)
print(lista1 == lista2)

False
True


In [48]:
lista3 = lista2
lista3 is lista2

True

Que sucede si modifico un elemento de la lista2 y veo la lista3

In [49]:
lista2[0] = "nuevo_elemento"

lista3

['nuevo_elemento', 10.5, 11.0, 11.5, 12.0, 12.5, 13.0, 13.5, 14.0, 14.5]

In [51]:
lista4 = lista1.copy()
lista1[0] = "nuevo2"
lista4

['nuevo2', 10.5, 11.0, 11.5, 12.0, 12.5, 13.0, 13.5, 14.0, 14.5]

## Dictionary Comprehension

Puedo crear diccionarios usando Dictionary Comprehension de la sigiente forma:

`dictionary = {key: value for vars in iterable}`

In [52]:
dic_comp ={k: v for k, v in [("v1", 1), ("v2", 2), ("v3", 3)]}
dic_comp

{'v1': 1, 'v2': 2, 'v3': 3}

Con un for loop:

In [57]:
cuadrado = dict()
for num in range(1, 11):
    cuadrado[num] = num*num
print(cuadrado)

{1: 1, 2: 4, 3: 9, 4: 16, 5: 25, 6: 36, 7: 49, 8: 64, 9: 81, 10: 100}


Usando Dictionary Comprehension

In [58]:
cuadrado = {num: num*num for num in range(1, 11)}
print(cuadrado)

{1: 1, 2: 4, 3: 9, 4: 16, 5: 25, 6: 36, 7: 49, 8: 64, 9: 81, 10: 100}


## Ejercicios 

1. Crear una lista con valores del 100 al 500 a intervaloes de 50


In [68]:
ej1 = []

for i in range(100,550,50):
    ej1.append(i)

ej1


[100, 150, 200, 250, 300, 350, 400, 450, 500]


2. Convertir esa lista en un diccionario donde los valores sean los elementos de la lista y las llaves el texto 'pago_{i}' donde i es el indice del pago, ej. pago_1 = 100

In [80]:
ej2 = dict()

for i in range(9):
    ej2[f"pago_{i}"] = ej1[i]

ej2

{'pago_0': 100,
 'pago_1': 150,
 'pago_2': 200,
 'pago_3': 250,
 'pago_4': 300,
 'pago_5': 350,
 'pago_6': 400,
 'pago_7': 450,
 'pago_8': 500}

3. Sacar el valor actual de la lista considerandola una lista de pagos a intervalos regulares con una tasa de interes del 100%. El valor actual es la suma de los valores individuales descontados por el interes, cada pago se descuenta dividiendolo por (1 + tasa)^indice, por ej. 150 / ((1 + 1) ** 2). Recordar que los indices empiezan en 0

In [84]:
ej3 = dict()
tasa = 1

for i in range(9):
    ej3[f"interes_{i}"] =  ej2.get(f"pago_{i}") / ((1 + tasa) ** (i+1))

ej3

{'interes_0': 50.0,
 'interes_1': 37.5,
 'interes_2': 25.0,
 'interes_3': 15.625,
 'interes_4': 9.375,
 'interes_5': 5.46875,
 'interes_6': 3.125,
 'interes_7': 1.7578125,
 'interes_8': 0.9765625}